File: api_client.py
import requests
from config import PROFILE_URL
from auth import get_access_token


def get_user_profile():
    token = get_access_token()

    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/json"
    }

    response = requests.get(PROFILE_URL, headers=headers)

    if response.status_code == 200:
        return response.json()

    return {
        "error": response.status_code,
        "message": response.text
    }


if __name__ == "__main__":
    profile = get_user_profile()
    print(profile)

File: milestone_4_code/frontend/src/services/apiClient.ts
import axios from "axios";
import { STORAGE_KEYS, API_BASE_URL } from "../constants";

/**
 * Pre-configured Axios instance for all API calls.
 * - Automatically attaches JWT bearer token from localStorage.
 * - Handles 401 responses by clearing stale tokens.
 */
export const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    "Content-Type": "application/json",
  },
});

// Request interceptor: attach auth token
apiClient.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem(STORAGE_KEYS.TOKEN);
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

// Response interceptor: handle 401 globally
apiClient.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response?.status === 401) {
      localStorage.removeItem(STORAGE_KEYS.TOKEN);
      // Only redirect if not already on login page
      if (!window.location.pathname.includes("/login")) {
        window.location.href = "/login";
      }
    }
    return Promise.reject(error);
  }
);
